src/Listeners/DetectSchemaConflicts.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+namespace Plank\Publisher\Listeners;
+
+use Closure;
+use Illuminate\Container\Container;
+use Illuminate\Database\Connection;
+use Illuminate\Database\Events\MigrationEnded;
+use Illuminate\Database\Schema\Builder;
+use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\Queue;
+use Plank\Publisher\Contracts\DetectsConflicts;
+use Plank\Publisher\Contracts\ResolvesKeys;
+use Plank\Publisher\Jobs\ResolveSchemaConflicts;
+
+class DetectSchemaConflicts
+{
+    public function __construct(
+        protected Container $container,
+    ) {}
+
+    public function handle(MigrationEnded $event)
+    {
+        /** @var ResolvesKeys $resolver */
+        $resolver = config()->get('publisher.conflicts.resolver');
+
+        /** @var ResolveSchemaConflicts $job */
+        $job = config()->get('publisher.conflicts.job');
+
+        $this->usingConflictSchema(fn () => $event->migration->{$event->method}())
+            ->each(function (Collection $conflicts) use ($resolver, $job) {
+                if ($conflicts->isEmpty()) {
+                    return;
+                }
+
+                $pk = $resolver::fromTable($conflicts->first()->table);
+
+                Queue::pushOn(
+                    config()->get('publisher.conflicts.queue'),
+                    new $job($pk, $conflicts),
+                );
+            });
+    }
+
+    /**
+     * @return Collection<Collection<Conflict>>
+     */
+    protected function usingConflictSchema(Closure $callback): Collection
+    {
+        $active = $this->container->make('db.schema');
+
+        /** @var Connection $db */
+        $db = $this->container->make('db.connection');
+
+        /** @var class-string<Builder&DetectsConflicts> $class */
+        $class = config()->get('publisher.conflicts.schema');
+        $schema = new $class($db);
+
+        $this->container->instance('db.schema', $schema);
+
+        $db->pretend(function () use ($callback) {
+            return $callback();
+        });
+
+        $this->container->instance('db.schema', $active);
```

Comment:
clever, so at event hook time, we check what the active schema builder is and swap it out for the `ConflictSchema`? I do wonder if this might be a source of a race condition of some kind, if something else is interacting with the schema builder at the same time. I don't _think_ thats the case, but i'm also not 100% sure

Reply:
A good thought, but I am pretty confident this should be 👌 

The application instance will not be doing anything else while this code is executing. The Event system in laravel is not async.

We can get @a-drew to weigh in here too, but to my knowledge the code will be running synchronously and be blocking, therefore there should be no ability to race.

Reply:
I don't think so either but let's check 🕵️ 

I suspect `instance` replaced active version from that moment onward. So unless something else is also trying to replace the active db schema, we shouldn't have any conflicts.

Reply:
The whole process can be "blackboxed" as a branch of the program flow. By retrieving, swapping and replacing the instance there should be no side effects. It is only our code that would be further swapping the schema builder instance.

Reply:
yeah, since its not async events and usually with the context of 1 single request, I don't see anything would be swapping out the schema instance in the middle of this listener. And even in a project with something like octane, that would be the http side and not affect cli commands that run these migrations and evoke the MigrationEnded event.